token-repository/src/main/scala/io/renku/tokenrepository/repository/init/TokensMigrator.scala

```diff
@@ -0,0 +1,159 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.renku.tokenrepository.repository
+package init
+
+import AccessTokenCrypto.EncryptedAccessToken
+import ProjectsTokensDB.SessionResource
+import association.TokenStoringInfo.Project
+import association._
+import cats.effect.{Async, Temporal}
+import cats.syntax.all._
+import deletion.TokenRemover
+import io.renku.db.DbClient
+import io.renku.graph.model.projects
+import io.renku.http.client.{AccessToken, GitLabClient}
+import io.renku.metrics.LabeledHistogram
+import org.typelevel.log4cats.Logger
+import skunk.~
+
+import scala.concurrent.duration._
+
+private object TokensMigrator {
+  def apply[F[_]: Async: GitLabClient: SessionResource: Logger](
+      queriesExecTimes: LabeledHistogram[F]
+  ): F[DBMigration[F]] = for {
+    accessTokenCrypto         <- AccessTokenCrypto[F]()
+    tokenValidator            <- TokenValidator[F]
+    tokenRemover              <- TokenRemover[F](queriesExecTimes).pure[F]
+    projectAccessTokenCreator <- ProjectAccessTokenCreator[F]()
+    associationPersister      <- AssociationPersister[F](queriesExecTimes).pure[F]
+  } yield new TokensMigrator[F](accessTokenCrypto,
+                                tokenValidator,
+                                tokenRemover,
+                                projectAccessTokenCreator,
+                                associationPersister,
+                                queriesExecTimes
+  )
+}
+
+private class TokensMigrator[F[_]: Async: SessionResource: Logger](
+    tokenCrypto:               AccessTokenCrypto[F],
+    tokenValidator:            TokenValidator[F],
+    tokenRemover:              TokenRemover[F],
+    projectAccessTokenCreator: ProjectAccessTokenCreator[F],
+    associationPersister:      AssociationPersister[F],
+    queriesExecTimes:          LabeledHistogram[F],
+    retryInterval:             Duration = 5 seconds
+) extends DbClient[F](Some(queriesExecTimes))
+    with DBMigration[F]
+    with TokenRepositoryTypeSerializers {
+
+  private val logPrefix = "token migration:"
+
+  import associationPersister._
+  import fs2.Stream
+  import io.renku.db.SqlStatement
+  import projectAccessTokenCreator._
+  import skunk.Void
+  import skunk.implicits._
+  import tokenCrypto._
+  import tokenRemover._
+  import tokenValidator._
+
+  override def run(): F[Unit] =
+    Stream
+      .emit(())
+      .repeat
+      .evalMap(_ => findTokenWithoutDates)
+      .takeThrough(_.nonEmpty)
+      .flatMap(maybeProjectAndToken => Stream.emits(maybeProjectAndToken.toList))
```

Comment:
I think this could be same as `.unNoneTerminate` (stopping on first `None`)? (there also is `.unNone` to just filter out all `None`s)

Reply:
Perfect! It's so nice to have fs2 skilled eye on the team ;)